tmail-backend/blob/blob-guice/src/main/java/com/linagora/tmail/blob/guice/BlobStoreModulesChooser.java

```diff
@@ -58,11 +40,75 @@
-    private static final String PRE_LAST_NAMED = "pre_last";
-    private static final String LAST_NAMED = "last";
+    public static final String FIRST_LEVEL = "first_level_blob_store_dao";
+    public static final String SECOND_LEVEL = "second_level_blob_store_dao";
+    public static final String THIRD_LEVEL = "third_level_blob_store_dao";
+    public static final String FOURTH_LEVEL = "fourth_level_blob_store_dao";
+    public static final String SECONDARY_BLOB_STORE_DAO = "secondary_blob_store_dao";
+    public static final String SECONDARY_S3_CLIENT_FACTORY = "secondary_s3_client_factory";
+
+    static class BaseObjectStorageModule extends AbstractModule {
+        @Override
+        protected void configure() {
+            install(new S3BucketModule());
+            install(new S3BlobStoreModule());
+            bind(BlobStoreDAO.class).annotatedWith(Names.named(FIRST_LEVEL))
+                .to(S3BlobStoreDAO.class);
+        }
+
+        @Provides
+        @Singleton
+        BlobStoreDAO providePrimaryBlobStoreDAO(@Named(FOURTH_LEVEL) BlobStoreDAO blobStoreDAO) {
+            return blobStoreDAO;
+        }
+    }
+
+    static class NoSecondaryObjectStorageModule extends AbstractModule {
+        @Provides
+        @Singleton
+        @Named(SECOND_LEVEL)
+        BlobStoreDAO provideBlobStoreDAO(@Named(FIRST_LEVEL) BlobStoreDAO blobStoreDAO) {
+            return blobStoreDAO;
+        }
+    }
+
+    static class SecondaryObjectStorageModule extends AbstractModule {
+        private S3BlobStoreConfiguration secondaryS3BlobStoreConfiguration;
+
+        public SecondaryObjectStorageModule(S3BlobStoreConfiguration secondaryS3BlobStoreConfiguration) {
+            this.secondaryS3BlobStoreConfiguration = secondaryS3BlobStoreConfiguration;
+        }
+
+        @Provides
+        @Singleton
+        @Named(SECONDARY_S3_CLIENT_FACTORY)
+        S3ClientFactory provideSecondaryS3ClientFactory(MetricFactory metricFactory,
+                                                       GaugeRegistry gaugeRegistry) {
+            return new S3ClientFactory(secondaryS3BlobStoreConfiguration, metricFactory, gaugeRegistry);
+        }
+
+        @Provides
+        @Singleton
+        @Named(SECONDARY_BLOB_STORE_DAO)
+        BlobStoreDAO provideSecondaryS3BlobStoreDAO(@Named(SECONDARY_S3_CLIENT_FACTORY) S3ClientFactory s3ClientFactory,
+                                                      BlobId.Factory blobIdFactory) {
+            return new S3BlobStoreDAO(s3ClientFactory, secondaryS3BlobStoreConfiguration, blobIdFactory);
+        }
+
+        @Provides
+        @Singleton
+        @Named(SECOND_LEVEL)
+        BlobStoreDAO provideBlobStoreDAO(@Named(FIRST_LEVEL) BlobStoreDAO firstBlobStoreDAO,
+                                        @Named(SECONDARY_BLOB_STORE_DAO) BlobStoreDAO secondBlobStoreDAO) {
+            return new SecondaryBlobStoreDAO(firstBlobStoreDAO, secondBlobStoreDAO);
+        }
+    }
```

Comment:
We don't need 
`SECONDARY_S3_CLIENT_FACTORY` 
`SECONDARY_BLOB_STORE_DAO`
when we don't reuse it anywhere. 

More guice name key make reading more difficult.
```suggestion
            static class SecondaryObjectStorageModule extends AbstractModule {
        private final S3BlobStoreConfiguration secondaryS3BlobStoreConfiguration;

        public SecondaryObjectStorageModule(S3BlobStoreConfiguration secondaryS3BlobStoreConfiguration) {
            this.secondaryS3BlobStoreConfiguration = secondaryS3BlobStoreConfiguration;
        }

        @Provides
        @Singleton
        @Named(SECOND_LEVEL)
        BlobStoreDAO provideBlobStoreDAO(@Named(INITIAL) BlobStoreDAO firstBlobStoreDAO,
                                         BlobId.Factory blobIdFactory,
                                         MetricFactory metricFactory,
                                         GaugeRegistry gaugeRegistry) {
            S3ClientFactory s3SecondaryClientFactory = new S3ClientFactory(secondaryS3BlobStoreConfiguration, metricFactory, gaugeRegistry);
            S3BlobStoreDAO secondaryBlobStoreDAO = new S3BlobStoreDAO(s3SecondaryClientFactory, secondaryS3BlobStoreConfiguration, blobIdFactory);
            return new SecondaryBlobStoreDAO(firstBlobStoreDAO, secondaryBlobStoreDAO);
        }
    }
```